api/src/main/java/org/hawkular/inventory/base/Mutator.java

```diff
@@ -0,0 +1,431 @@
+            } else {
+                deleted.add(toDelete);
+            }
+            deletedRels.addAll(context.backend.getRelationships(toDelete, both));
+
+            //we've gathered all entities to be deleted. Now convert them all to entities for reporting purposes.
+            //We have to do it prior to actually deleting the objects in the backend so that all information and
+            //relationships is still available.
+            deletedEntities = deleted.stream().map((o) -> context.backend.convert(o, context.backend.extractType(o)))
+                    .collect(Collectors.<AbstractElement<?, ?>>toSet());
+
+            deletedRelationships = deletedRels.stream().map((o) -> context.backend.convert(o, Relationship.class))
+                    .collect(toSet());
+
+            //k, now we can delete them all... the order is not important anymore
+            for (BE e : deleted) {
+                context.backend.delete(e);
+            }
+
+            for (BE e : verticesToDeleteThatDefineSomething) {
+                if (context.backend.hasRelationship(e, outgoing, defines.name())) {
+                    //we avoid the convert() function here because it assumes the containing entities of the passed in
+                    //entity exist. This might not be true during the delete because the transitive closure "walks" the
+                    //entities from the "top" down the containment chain and the entities are immediately deleted.
+                    String rootId = context.backend.extractId(toDelete);
+                    String definingId = context.backend.extractId(e);
+                    String rootType = context.entityClass.getSimpleName();
+                    String definingType = context.backend.extractType(e).getSimpleName();
+
+                    String rootEntity = "Entity[id=" + rootId + ", type=" + rootType + "]";
+                    String definingEntity = "Entity[id=" + definingId + ", type=" + definingType + "]";
+
+                    throw new IllegalArgumentException("Could not delete entity " + rootEntity + ". The entity " +
+                            definingEntity + ", which it (indirectly) contains, acts as a definition for some " +
+                            "entities that are not deleted along with it, which would leave them without a " +
+                            "definition. This is illegal.");
+                } else {
+                    context.backend.delete(e);
+                }
+            }
+
+            context.backend.commit();
+        } catch (Exception e) {
+            context.backend.rollback();
+            throw e;
+        }
+
+        //report the relationship deletions first - it would be strange to report deletion of a relationship after
+        //reporting that an entity on one end of the relationship has been deleted
+        for (Relationship r : deletedRelationships) {
+            context.notify(r, deleted());
```

Comment:
Isn't entity-based notification enough? Perhaps adding/removing an edge can be considered also as the entity update, I don't know.